### DISCUSSION TITLE
How to read file record (0x14)?

### DISCUSSION THREAD

**Discussion by klgill24** [pymodbus-dev/pymodbus#2404] [Q&A] [answered]
I missed some example for this. But i tried something. This gives me successfully that what i want in the logging. But there ist no hint to use the result. Where is the file?

from pymodbus.constants import Endian
from pymodbus.payload import BinaryPayloadDecoder
from pymodbus.payload import BinaryPayloadBuilder
from pymodbus.client import ModbusTcpClient
from pymodbus.pdu.file_message import FileRecord, ReadFileRecordRequest
import logging

logging.basicConfig()
log = logging.getLogger()
log.setLevel(logging.DEBUG)

client = ModbusTcpClient('192.168.178.27')
client.connect()
records = []
record1 = FileRecord(reference_type=0x06, file_number=0x5a, record_number=0x01, record_length=0x0e)
records.append(record1)
request = ReadFileRecordRequest(records=records)
response = client.execute(request)
#decoder = BinaryPayloadDecoder.fromRegisters(result, byteorder=Endian.BIG, wordorder=Endian.BIG)
if not response.isError():
    print(response.records)
    print(response)
else:
    # Handle Error
    print(response)
#print (decoder.decode_32bit_float())
client.close()

logging:
DEBUG:pymodbus.logging:Connection to Modbus server established. Socket ('192.168.178.24', 59422)
DEBUG:pymodbus.logging:Current transaction state - IDLE
DEBUG:pymodbus.logging:Running transaction 1
DEBUG:pymodbus.logging:SEND: 0x0 0x1 0x0 0x0 0x0 0xa 0x1 0x14 0x7 0x6 0x0 0x5a 0x0 0x1 0x0 0xe
DEBUG:pymodbus.logging:New Transaction state "SENDING"
DEBUG:pymodbus.logging:Changing transaction state from "SENDING" to "WAITING FOR REPLY"
DEBUG:pymodbus.logging:Changing transaction state from "WAITING FOR REPLY" to "PROCESSING REPLY"
DEBUG:pymodbus.logging:RECV: 0x0 0x1 0x0 0x0 0x0 0x21 0x1 0x14 0x1e 0x1d 0x6 0x0 0x0 0x0 0x0 0x67 0x17 0xd4 0x29 0x42 0x98 0x87 0xfc 0x0 0x0 0x0 0x0 0x67 0x16 0xce 0xe0 0x42 0xc3 0x6b 0x71 0x0 0x0 0x0 0x0
DEBUG:pymodbus.logging:Processing: 0x0 0x1 0x0 0x0 0x0 0x21 0x1 0x14 0x1e 0x1d 0x6 0x0 0x0 0x0 0x0 0x67 0x17 0xd4 0x29 0x42 0x98 0x87 0xfc 0x0 0x0 0x0 0x0 0x67 0x16 0xce 0xe0 0x42 0xc3 0x6b 0x71 0x0 0x0 0x0 0x0
DEBUG:pymodbus.logging:Factory Response[ReadFileRecordResponse': 20]
DEBUG:pymodbus.logging:Frame advanced, resetting header!!
DEBUG:pymodbus.logging:Adding transaction 1
DEBUG:pymodbus.logging:Getting transaction 1
DEBUG:pymodbus.logging:Changing transaction state from "PROCESSING REPLY" to "TRANSACTION_COMPLETE"
[FileRecord(file=0, record=0, length=14)]
<pymodbus.pdu.file_message.ReadFileRecordResponse object at 0x7fff27dc7b90>

**Comment by janiversen** [comment]
?? you send a request for a record of the file, not the whole file.

Please read the modbus standard, it explains how to read files.

The record is delivered back in response, but I wonder why you use execute instead of the API available.

**Reply by klgill24** [reply]
Its because of a missing example. The modbus standard does not really help in this way.
In Processing there is exactly the result that i want.

**Reply by janiversen** [reply]
I was just trying to say you could simplify your program by using the def read_file_record(... call instead of building the du yourself, but of course it is your choice.

The modbus standard shows exactly what is returned from the requests, and in return from execute you have the response object. I am not sure what more you want, you can simply save it to disk if you want a normal file.

Please remember, which you know if having read the standard, a modbus file is not the same as a file on your disk.

**Reply by janiversen** [reply]
There are no example because it would simply be a loop over read_file_record()

**Comment by janiversen** [comment]
Also I am puzzled why do you use a FileRecord as parameter to the request pdu, that is not correct (FileRecord is in the response object)

You might not want to use the simpler call file_read_record, which is a shame, because it actually defines the parameters needed.

**Reply by klgill24** [reply]
Its not a shame. Its simply a non knowing. I unterstand your answer. For reading registers there are much examples right here in the forum. But there is no example for doing the file read. It was looking everywhere for a howto, even in the library itself. I just guess how to do it. Even i have no clue to decode the response or print it.
For you it must be 2 minutes to give a working example.

**Reply by janiversen** [reply, marked as answer]
I think the problem is not the example, but more what you want to achieve, normally apps only read the records they are interested in, typically it does not make sense to read all records, and even less to make a real file.

You handle the FileReadRecordRequest class, so what is the problem in handling the FileReadRecordRespomse class, which is what you get back.

Not knowing, sounds like you have not read our documentation, there all calls are documented with parameters.

The content of the file records are highly vendor dependent, so an example would only be able to print the hex information, as we already do in the log,

**Reply by klgill24** [reply]
I have read the documentation, but handling of the FileReadRecordResponse i have no clue. Even printing the hex information would help. I handled the FileReadRecordRequest because there was an wrong example in Stackoverflow from 2017.
Why its so difficult to give two lines of code.

**Reply by janiversen** [reply]
It is a lot more than 2 lines of code, and they will not help you further, because you already get a response returned, you just need to use it...and that part is vendor dependent so not something I can help with.

**Reply by janiversen** [reply]
With the response stops the responsibility of pymodbus, how you use the data is not part of pymodbus.

An example would just be the code you already have, cleaned a bit...how you use the data returned is something you need to read in the manual of your device. As I said before a modbus file is just a sequence of records, not a physical file.

**Reply by klgill24** [reply]
This is exactly what i mean. I know to use the data. But i don't know how to get the data from the response. 

This is simple:
client = ModbusTcpClient('192.168.178.27')
client.connect()
request = client.read_holding_registers(1,2,1)
result = request.registers
decoder = BinaryPayloadDecoder.fromRegisters(result, byteorder=Endian.BIG, wordorder=Endian.BIG)
print (decoder.decode_32bit_float())
client.close()

But i can't use with File Record
response = client.execute(request)
result=response.registers
or 
result=response.encode()
or 
result=response.records

where is the list or array with the values

**Reply by janiversen** [reply]
Did you bother to examine the FileReadRecordResponse class, that is the response NOT FileRecord ! 

You actually pointed at the list "response.records", but that is only part of the response....there are NO list of values !!

"response.records" might be a list of FileRecord, but it depends totally on your device...f.x. one of my devices return an int for each record.

Run your program with a debugger, then you can see:
a) the name of class which variable response contains
b) the members of the class (variables), which each have a specific meaning.

The "variables" in the modbus response, is simply transformed into python object variables, and available to the app.

**Comment by janiversen** [comment]
Please remember your device might also respond with no records, which is stored in another object variable....you really need to see FileReadRecordResponse with the corresponding page in the modbus standard.

you write:

This is simple:
client = ModbusTcpClient('192.168.178.27')
client.connect()
request = client.read_holding_registers(1,2,1)
result = request.registers
decoder = BinaryPayloadDecoder.fromRegisters(result, byteorder=Endian.BIG, wordorder=Endian.BIG)
print (decoder.decode_32bit_float())
client.close()

but it seem you made it a lot less simple, calling the BinaryPayloadDecodder is not correct, and on some CPUs very wrong.

**Comment by klgill24** [comment]
Thanks, debugging does the thing. I have the values.

**Comment by klgill24** [comment]
Maybe this or something similar can go to the examples. It is for Sentron Pac2200:

from pymodbus.constants import Endian
from pymodbus.client import ModbusTcpClient
from pymodbus.pdu.file_message import FileRecord, ReadFileRecordRequest
from pymodbus.payload import BinaryPayloadDecoder
import datetime

client = ModbusTcpClient('IP-Adress')
client.connect()
pac_history_start = 3
pac_history_record_length = 6
pac_history_records = 15
rec_pos=pac_history_start
while rec_pos < pac_history_start+pac_history_record_length*pac_history_records:
	records = [FileRecord(reference_type=0x06, file_number=90, record_number=rec_pos, record_length=pac_history_record_length)]
	response = client.read_file_record(records)
	rec_pos=rec_pos+pac_history_record_length
	if not response.isError():
		result=response.records[0].record_data
		decoder = BinaryPayloadDecoder(result, byteorder=Endian.BIG, wordorder=Endian.BIG)
		date_time=datetime.datetime.fromtimestamp(decoder.decode_32bit_uint())
		hist_val1=decoder.decode_32bit_float()
		hist_val2=decoder.decode_32bit_float()
		print(str(hist_val1)+" "+str(hist_val2)+" "+date_time.strftime('%d.%m.%Y %H:%M:%S'))
		decoder.reset()
	else:
    	# Handle Error
		print(response)
client.close()

**Reply by janiversen** [reply]
feel free to submit a pull requests that adds it to examples/contrib, so that it is available for others.
